baseapp/abci.go

```diff
@@ -238,19 +241,52 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
 // work in a block before proposing it.
 //
 // Transactions can be modified, removed, or added by the application. Since the
-// application maintains it's own local mempool, it will ignore the transactions
+// application maintains its own local mempool, it will ignore the transactions
 // provided to it in RequestPrepareProposal. Instead, it will determine which
 // transactions to return based on the mempool's semantics and the MaxTxBytes
 // provided by the client's request.
 //
-// Note, there is no need to execute the transactions for validity as they have
-// already passed CheckTx.
-//
 // Ref: https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-060-abci-1.0.md
 // Ref: https://github.com/tendermint/tendermint/blob/main/spec/abci/abci%2B%2B_basic_concepts.md
 func (app *BaseApp) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
-	// TODO: Implement.
-	return abci.ResponsePrepareProposal{Txs: req.Txs}
+	var (
+		txsBytes  [][]byte
+		byteCount int64
+	)
+
+	ctx := app.getContextForTx(runTxPrepareProposal, []byte{})
+	iterator := app.mempool.Select(ctx, req.Txs)
+
+	for iterator != nil {
+		memTx := iterator.Tx()
+
+		bz, err := app.txEncoder(memTx)
+		if err != nil {
+			panic(err)
+		}
+
+		txSize := int64(len(bz))
+
+		// NOTE: runTx was already run in CheckTx which calls mempool.Insert so ideally everything in the pool
+		// should be valid. But some mempool implementations may insert invalid txs, so we check again.
+		_, _, _, _, err = app.runTx(runTxPrepareProposal, bz)
```

Comment:
I don't necessarily understand the point of forcing the `runTx` behavior on all app-side mempool implementations. If anything, this just makes the `SetMempool` API more restrictive. It also forces the performance burden on all custom app-side mempools of running all transactions through the Antehandlers a third time (`CheckTx`, `DeliverTx`, and now `PrepareProposal`). As stated in the comment here this shouldn't be necessary unless the application developer's custom app-side mempool implementation has a bug or is trying to do something unexpected. They can already break themselves in other ways, for example a developer could add some `panic` to their `Select` implementation which means no proposal would ever be sent, right?

Just a suggestion and not a dealbreaker, but I think I'd prefer to be able to opt out of this somehow.

Reply:
I disagree that they add _no_ performance burden. Some of the checks read from state which could potentially mean disk reads if the state is not in cache. There is also signature verification which is not free.

A couple of examples of reading from state during Antehandlers:
* Reading the Account in `ConsumeTxSizeGasDecorator` [here](https://github.com/cosmos/cosmos-sdk/blob/a2761bdc6f23500a2b38d398ee2a423f2a5311c0/x/auth/ante/basic.go#L114).
* Calling `SendCoinsFromAccountToModule` in `DeductFees` which reads module accounts from state [here](https://github.com/cosmos/cosmos-sdk/blob/a2761bdc6f23500a2b38d398ee2a423f2a5311c0/x/auth/ante/fee.go#L134).

There is also signature verifications [here](https://github.com/cosmos/cosmos-sdk/blob/a2761bdc6f23500a2b38d398ee2a423f2a5311c0/x/auth/ante/sigverify.go#L286) which looks to take ~0.15ms per-op for `secp256k1` if I am understanding the [benchmarks](https://dashboard.bencher.orijtech.com/benchmark/f897726b91414429a6217146c28e821e) correctly (`Sig/secp256k1-8`). This means while ABCI is globally locked during `PrepareProposal`, you're doing this operation for each transaction serially. So if the block has 100 transactions on average, you're spending ~15ms on average per-block blocking the entire application and verifying transaction signatures that have already been verified.

Let me know if I'm misunderstanding something!

Reply:
> I disagree that they add no performance burden. Some of the checks read from state which could potentially mean disk reads if the state is not in cache. There is also signature verification which is not free.

True, state is read from but not written to. In the grand scheme of things, the performance overhead is near negligible.

Reply:
Again, we validate the txs (i.e. AnteHandler) since applications can insert txs on `Insert` during `CheckTx` and we need to ensure these are still valid when selecting during `PrepareProposal`.

Also, recall:

1. The application defines the AnteHandler chain
2. The application and thus it's AnteHandler is aware of tx execution context (i.e. `PrepareProposal`)

So if this doesn't suite your needs, your AnteHandler chain can easily do something like "If PrepareProposal, skip XYZ".

Reply:
> The application defines the AnteHandler chain
> So if this doesn't suite your needs, your AnteHandler chain can easily do something like "If PrepareProposal, skip XYZ".

This is a really good point that I had forgotten. I think this would be reasonable solution.